ipn/ipnext/ipnext.go

```diff
@@ -0,0 +1,284 @@
+	"errors"
+	"fmt"
+
+	"tailscale.com/control/controlclient"
+	"tailscale.com/ipn"
+	"tailscale.com/ipn/ipnauth"
+	"tailscale.com/tsd"
+	"tailscale.com/types/logger"
+	"tailscale.com/types/views"
+	"tailscale.com/util/mak"
+)
+
+// Extension augments LocalBackend with additional functionality.
+//
+// An extension uses the provided [Host] to register callbacks
+// and interact with the backend in a controlled, well-defined
+// and thread-safe manner.
+//
+// Extensions are registered using [RegisterExtension].
+//
+// They must be safe for concurrent use.
+type Extension interface {
+	// Name is a unique name of the extension.
+	// It must be the same as the name used to register the extension.
+	Name() string
+
+	// Init is called to initialize the extension when LocalBackend is initialized.
+	// If the extension cannot be initialized, it must return an error,
+	// and its Shutdown method will not be called on the host's shutdown.
+	// Returned errors are not fatal; they are used for logging.
+	// A [SkipExtension] error indicates an intentional decision rather than a failure.
+	Init(Host) error
+
+	// Shutdown is called when LocalBackend is shutting down,
+	// provided the extension was initialized. For multiple extensions,
+	// Shutdown is called in the reverse order of Init.
+	// Returned errors are not fatal; they are used for logging.
+	Shutdown() error
+}
+
+// NewExtensionFn is a function that instantiates an [Extension].
+// If a registered extension cannot be instantiated, the function must return an error.
+// If the extension should be skipped at runtime, it must return either [SkipExtension]
+// or a wrapped [SkipExtension]. Any other error returned is fatal and will prevent
+// the LocalBackend from starting.
+type NewExtensionFn func(logger.Logf, *tsd.System) (Extension, error)
+
+// SkipExtension is an error returned by [NewExtensionFn] to indicate that the extension
+// should be skipped rather than prevent the LocalBackend from starting.
+//
+// Skipping an extension should be reserved for cases where the extension is not supported
+// on the current platform or configuration, or depends on a feature that is not available,
+// or otherwise should be disabled permanently rather than temporarily.
+//
+// Specifically, it must not be returned if the extension is not required right now
+// based on user preferences, policy settings, the current tailnet, or other factors
+// that may change throughout the LocalBackend's lifetime.
+var SkipExtension = errors.New("skipping extension")
+
+// Definition describes a registered [Extension].
+type Definition struct {
+	name  string         // name under which the extension is registered
+	newFn NewExtensionFn // function that creates a new instance of the extension
+}
+
+// Name returns the name of the extension.
+func (d *Definition) Name() string {
+	return d.name
+}
+
+// Make instantiates the extension.
+func (d *Definition) Make(logf logger.Logf, sys *tsd.System) (Extension, error) {
```

Comment:
Not blocking on this: Would it be more readable to change this method to be named `MakeExtension`?